### DISCUSSION TITLE
Improve query performance

### DISCUSSION THREAD

**Discussion by vigneshshettyin** [ClickHouse/ClickHouse#54728] [Q&A]
Hi 

Below is my use case

Table structure

```
CREATE TABLE test_table
(
    `id` Int64,
    `code` LowCardinality(String),
    `cmo_id` LowCardinality(String),
    `label` String,
    `roll_up_date` Date,
    `total_clp` Int16,
    `total_plp` Int16,
)
ENGINE = MergeTree
PARTITION BY roll_up_date
ORDER BY (cmo_id, code, roll_up_date)
SETTINGS index_granularity = 8192  
```

I am running a single node Clickhouse Database on AWS (r5n.4xlarge). I have inserted almost close to 4 Billion records to this table my use case involves to run a query something like 
`
WITH locations AS (
    SELECT
        cmo_id,
        sum(total_clp) as total_clp,
        sum(total_plp) as total_plp
    FROM
        test_table
    where
        roll_up_date BETWEEN '2020-01-01'
        and '2023-12-31'
        and code in (
            'TMP8497',
            'LP01TMP18',
            'KLP1TMP2',
            'UIP01TMP3',
            'TMP91A',
            'TMP91A2',
            'TMP91Z0',
            'TMP91Z1',
            'TMP91Z',
            'TMP91Z2',
            'TMP9210',
            'TMP9200'
        )
    group by
        cmo_id
)
SELECT
    *
FROM
    listing_data
    left outer join locations ON listing_data.cmo_id = locations.cmo_id
order by
    total_clp desc
limit
    100 offset 10;
`

The above query takes around 30-40 Seconds. I have tried partition based on roll_up_date also tried projections but didn't observe any improvements. I am in need for a solution to improve  query performance.

Please do ask me if you need any more requirements from my end.

Thanks!!

**Comment by den-crane** [comment]
How fast is this query and what it returns?

```sql
select count()
(
SELECT cmo_id, sum(total_clp) as total_clp, sum(total_plp) as total_plp FROM test_table where roll_up_date BETWEEN '2020-01-01' and '2023-12-31' and 
  code in ( 'TMP8497', 'LP01TMP18', 'KLP1TMP2', 'UIP01TMP3', 'TMP91A', 'TMP91A2', 'TMP91Z0', 'TMP91Z1', 'TMP91Z', 'TMP91Z2', 'TMP9210', 'TMP9200' ) 
group by cmo_id 
)
```

How many rows is in `listing_data` ?

**Reply by den-crane** [reply]
I would partition by month `PARTITION BY toYYYYMM(roll_up_date)`

**Reply by vigneshshettyin** [reply]
Hey @den-crane 

Thanks for your quick response.

The query you mentioned takes around 30-40 seconds as it has almost 3 billion records it's scanning the whole table. Listing_data table has around 2.5 million records.

**Reply by den-crane** [reply]
>it has almost 3 billion records

`select count() ( SELECT cmo_id, sum(total_clp) as total_clp` ~ 3 billion records ?

**Reply by vigneshshettyin** [reply]
> >it has almost 3 billion records
> 
> 
> 
> `select count() ( SELECT cmo_id, sum(total_clp) as total_clp` ~ 3 billion records ?

No, almost 3 million records.

**Reply by vigneshshettyin** [reply]
We have around 2.5 million unique cmo_id's and around 0.1 million unique codes, each cmo_id has entry for 0.1 million unique codes.

**Reply by vigneshshettyin** [reply]
Hey, @den-crane Do you have any inputs to improve query performance?

**Reply by den-crane** [reply]
try

```sql
alter table test_table add projection test_table_sum
 (
    SELECT code, roll_up_date, cmo_id, sum(total_clp) as total_clp, sum(total_plp) as total_plp 
    group by code, roll_up_date, cmo_id
 );

alter table test_table materialize projection test_table_sum settings mutations_sync=2;
```

**Reply by vigneshshettyin** [reply]
Hey, @den-crane thanks for your response. 

`alter table test_table materialize projection test_table_sum settings mutations_sync=2;
`
The above query is timing out. Any solution to avoid this issue? 

_Timeout exceeded while receiving data from server. Waited for 300 seconds, timeout is 300 seconds.
Cancelling query_

**Reply by den-crane** [reply]
>Timeout exceeded while receiving data from server. Waited for 300 seconds, timeout is 300 seconds.

This error is irrelevant. This is a mutation. And will be finished any way. Check `select * from system.mutations where table = 'test_table' order by mutation_id desc limit 3`

**Reply by den-crane** [reply]
try your query again

**Comment by vigneshshettyin** [comment]
Yes, you are right I ran it again it completed. Also checked system.mutation table entry is_done flag was true for test_table_sum projection. Now the query response time as improved a lot. 

Query returns response within 1 second now

**Comment by vigneshshettyin** [comment]
Can you suggest anymore improvisations, one that I think of is current partition is for roll_up_date `(PARTITION BY roll_up_date).` As you suggest I will change it to `PARTITION BY toYYYYMM(roll_up_date)`

**Comment by vigneshshettyin** [comment]
Hi, @den-crane tried the same with 6 Billion rows dataset, query isn't performing well. What would be your suggestion to optimize to receive better query performance?

**Comment by vigneshshettyin** [comment]
My current table structure is

```
CREATE TABLE test_table
(
    `id` Int64,
    `code` LowCardinality(String),
    `cmo_id` LowCardinality(String),
    `label` String,
    `roll_up_date` Date,
    `clp_id` Int16,
    `plp_id` Int16,
)
ENGINE = MergeTree
PARTITION BY toYYYYMM(roll_up_date)
ORDER BY (cmo_id, code, roll_up_date)
SETTINGS index_granularity = 8192
```

I have also created projection for the same.

```
alter table test.test_table add projection test_projection
 (
    SELECT code, roll_up_date, cmo_id, count(distinct clp_id) as clp_id, count(distinct plp_id) as plp_id
    group by code, roll_up_date, cmo_id
 );
```

The query with which I am facing poor response time is

```
SELECT
    cmo_id,
    count(distinct clp_id) as clp_id,
    count(distinct plp_id) as plp_id
from
    test.test_table
where
    code in (
        '785P10',
        'POL0000',
        'IOPL0129',
        'Z79IKL',
        'FLKMD',
        'HJYIWSXD',
        'FIOKDS',
        '7990WJ',
        'LPOXEIO',
        'IISNEW'
    )
    and roll_up_date BETWEEN '2020-01-01'
    and '2023-12-31'
group by
    cmo_id
order by
    count(distinct clp_id) desc
limit
    10
```

This query performs low because we have maximum records for the codes mentioned in the query, due to this scanning table is taking time.  

Query execution 

`
10 rows in set. Elapsed: 59.698 sec. Processed 72.49 million rows, 28.26 GB (1.83 million rows/s., 711.78 MB/s.)
Peak memory usage: 41.82 GiB.
`

Please ask for any more information, Could anyone please suggest a optimized approach to improve the query performance

**Comment by melvynator** [comment]
Hello,

Reviewing your order by and your query it seems that you have an issue with your keys selection, do you have additional query that query cmo_id? If not then I will suggest to remove it from your orderby and to start with code. 

Also what the range of date you are querying? What's the smallest granularity? Weekly?  Daily? 

It would be interesting to run this query:

EXPLAIN indexes=1 YOUR_QUERY and look at how the primary index filters the granules and part that are being read? 

You can also test different join algorithm to speed-up your queries eventually: https://clickhouse.com/docs/en/operations/settings/settings#join_algorithm

**Reply by vigneshshettyin** [reply]
Hey, @melvynator I had done a typo while in the query I have rectified it. 
 The query result for EXPLAIN INDEX=1 QUERY
`┌─explain──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Expression ((Projection + Before ORDER BY))                                                                                      │
│   Limit (preliminary LIMIT (without OFFSET))                                                                                     │
│     Aggregating                                                                                                                  │
│       Filter                                                                                                                     │
│         ReadFromMergeTree (test_projection_hash)                                                                                 │
│         Indexes:                                                                                                                 │
│           PrimaryKey                                                                                                             │
│             Keys:                                                                                                                │
│               CODE_HASH                                                                                                          │
│               ROLL_UP_DATE                                                                                                       │
│             Condition: and((ROLL_UP_DATE in (-Inf, 19722]), and((ROLL_UP_DATE in [18262, +Inf)), (CODE_HASH in 11-element set))) │
│             Parts: 635/803                                                                                                       │
│             Granules: 5075/1446458                                                                                               │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘`

**Comment by den-crane** [comment]
>count(distinct claim_id) desc

I don't see `claim_id` column.

----

`count(distinct` is kinda slow.

Try this:

```
alter table test.test_table add projection test_projection
 (
    SELECT code, roll_up_date, cmo_id, uniq(lp_id) as clp_id, uniq(plp_id) as plp_id
    group by code, roll_up_date, cmo_id
 );

SELECT
    cmo_id,
    uniq(lp_id) as clp_id, 
    uniq(plp_id) as plp_id
from
    test.test_table
where
    code in (
        '785P10',
        'POL0000',
        'IOPL0129',
        'Z79IKL',
        'FLKMD',
        'HJYIWSXD',
        'FIOKDS',
        '7990WJ',
        'LPOXEIO',
        'IISNEW'
    )
    and roll_up_date BETWEEN '2020-01-01'
    and '2023-12-31'
group by
    cmo_id
limit
    10
```

**Reply by vigneshshettyin** [reply]
Hey, @den-crane 
Query performance has still not improved, with random 50 codes, query performs well it return results within 2-4 seconds. But when I increase the codes to like 100-200 the query response time is very bad it's around 217 s.

Ig it's because for the codes I passed table has around 736.90 million rows processing it is taking time. What are your inputs?

```
10 rows in set. Elapsed: 217.912 sec. Processed 736.90 million rows, 133.99 GB (3.38 million rows/s., 614.86 MB/s.)
Peak memory usage: 92.21 GiB.
```

**Reply by den-crane** [reply]
**I don't see claim_id column.**

try without order by. Probably because of claim_id

**Reply by vigneshshettyin** [reply]
That was a typo, tried without it only.

**Reply by den-crane** [reply]
Do you see the difference in the performance of the uniq with and without the projection?

**Reply by vigneshshettyin** [reply]
Yes, it's better then

```
    count(distinct clp_id) as clp_id,
    count(distinct plp_id) as plp_id
```

**Reply by den-crane** [reply]
Sorry, I asked another question. Do you see the difference in the performance of the uniq **with** and **without** the projection? I am trying to understand does projection help or not.

**Reply by vigneshshettyin** [reply]
Yes, yes without Projection query performance is very bad.

**Reply by vigneshshettyin** [reply]
Hi, @den-crane what is your view on applying set or bloom_filter index on the table for cmo_id, roll_up_date and code instead of using Projection?

**Reply by vigneshshettyin** [reply]
If we use Projection the iteration/rows is reducing 

`10 rows in set. Elapsed: 59.698 sec. Processed 72.49 million rows, 28.26 GB (1.83 million rows/s., 711.78 MB/s.) Peak memory usage: 41.82 GiB.`

Without Projection 
```

10 rows in set. Elapsed: 257.543 sec. Processed 11.70 billion rows, 256.42 GB (246.04 million rows/s., 5.39 GB/s.)
Peak memory usage: 2.74 GiB.
```

**Comment by MacNale** [comment]
@ vigneshshettyin Did you manage to resolve the performance problem?  Can you share the solution you decided to go with?

**Reply by vigneshshettyin** [reply]
No @MacNale, still looking for query optimization :(

**Comment by vigneshshettyin** [comment]
Hey, @den-crane here going with uniqExact will not be a great option for me ig as the memory usage is high for the query. If I go with Approximate result for uniq, uniqTheta is there anyway to roll up the table to reduce the no of rows scanned?

**Reply by den-crane** [reply]
we discussed it, projections and mat views/summingmergeTree/aggregatingMergeTree allow to roll up the table to reduce the no of rows scanned

**Reply by vigneshshettyin** [reply]
Mat view def

```
CREATE MATERIALIZED VIEW test.my_uniq_count_view
ENGINE = AggregatingMergeTree
ORDER BY CMO_ID, month, CODE
POPULATE AS
	SELECT 
		CMO_ID, 
		toStartOfMonth(ROLL_UP_DATE) AS month,
                CODE,
		uniqState(CLP_ID) AS cpl_id, 
		uniqState(PLP_ID) AS plp_id 
	FROM test.test_table 
	group by CMO_ID, month, CODE;
```

Mat view query

```
SELECT
	CMO_ID,
	uniqMerge(uniq_clp),
	uniqMerge(uniq_plp)
FROM test.my_uni_count_view
WHERE month = <whatever>
AND CODE = <whatever>
GROUP BY CMO_ID;
```

Can  you suggest any improvisation for this mat view also how can I create it using Distributed table? @den-crane

**Reply by vigneshshettyin** [reply]
Here the query is taking around 80s any optimization can you suggest?

Current infra 
2 shards of 16vCPU & 128 GB Memory 

@den-crane
